### PR TITLE
WEBUI-2189: Return user to originating page after session-expiry re-login [maintenance-3.1.x]

### PR DESCRIPTION
**Pull request by narasimhahyland**
## Problem
After WEBUI-1987 shipped, an inactivity/401 session expiry now redirects straight to the timeout login page and, after the user re-authenticates, drops them on the app root (dashboard) instead of the page they were viewing. The originating page context is lost. Jira: [WEBUI-2189](https://hyland.atlassian.net/browse/WEBUI-2189)

## Root cause
`NuxeoInactivityBehavior._logoutRedirect()` (`elements/behaviors/nuxeo-inactivity-behavior.js`) ends the server session and navigates to `login.jsp?nxtimeout=true` with no record of where the user was. It deliberately bypasses the SPA anonymous-login bounce (which nests/drops a `requestedUrl`), so the server never learns the origin page. `location.replace()` also removes it from history, so the back button can't recover it.

## Changes
* **`elements/behaviors/nuxeo-inactivity-behavior.js`**: before the logout navigation, `_saveRequestedUrl()` persists the current URL to `sessionStorage` (only on the inactivity/401 path — manual Sign Out is unaffected). New `_restoreRequestedUrlAfterLogin()` consumes it once on the next boot and navigates back **only** when the saved URL is same-origin (open-redirect guard using `origin + '/'`, not a bare prefix) and differs from the current page.
* **`elements/nuxeo-app.js`**: call `_restoreRequestedUrlAfterLogin()` at the start of `ready()`, before the timer/401 handlers are armed.
* **`test/nuxeo-inactivity-behavior.test.js`**: 9 new tests (save-on-logout, same-origin restore + consume-once, cross-origin/look-alike-host rejection, current-page no-op, empty no-op, sessionStorage-unavailable handling).

Why `sessionStorage` (not the ticket's suggested `localStorage`): it survives the same-tab logout → login → reload sequence but is scoped to the originating tab and cleared when the tab closes — bounding staleness and preventing a cross-tab hijack — while remaining same-origin.

## Test plan
- [x] `npm run lint` passes
- [x] `npm test` passes (2401 passed, 0 failed on lts-2025; identical code cherry-picked here)
- [ ] Manual: set `session.timeout` low, open a document, go idle → land on `login.jsp` timeout message → log back in → verify you return to the document (not the dashboard). Also verify manual Sign Out still lands on login/home.

## Notes
Backport pair of the `lts-2025` PR (#3421); same commit cherry-picked. Follow-up to WEBUI-1987 (both bases already have the merged behavior).

Made with [Cursor](https://cursor.com)

[WEBUI-2189]: https://hyland.atlassian.net/browse/WEBUI-2189?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by AnilKumarVanga — DISMISSED** [review]

File: elements/behaviors/nuxeo-inactivity-behavior.js [hunk 1]
```diff
@@ -272,6 +283,74 @@ export const NuxeoInactivityBehavior = {
...
+    // Restrict to the UI base path (e.g. `/nuxeo/ui/`, which implies same-origin) so a saved value can't
+    // slip past into an open redirect. baseUrl is absent in tests/edge cases → fall back to a bare '/'.
+    // Normalize to a single trailing '/' (mirroring the origin `+ '/'` reasoning): without it a baseUrl of
+    // `/nuxeo/ui` would let a look-alike sibling path such as `${origin}/nuxeo/ui-foo/...` pass startsWith().
+    const baseUrl = this.baseUrl || '/';
+    const normalizedBase = baseUrl.endsWith('/') ? baseUrl : `${baseUrl}/`;
+    const base = `${globalThis.location.origin}${normalizedBase}`;
+    if (requestedUrl?.startsWith(base) && requestedUrl !== globalThis.location.href) {
+      this._redirect(requestedUrl);
+    }
```
[the remark this answers is not in the document]
Comment (narasimhahyland) [reply elements/behaviors/nuxeo-inactivity-behavior.js]: Both valid — fixed in maintenance-3.1.x `42a70800e` (and lts-2025 #3421 as `592ec930f`).

- **Absolute baseUrl:** the base and the saved URL are now resolved with `new URL(base, location.origin)` / `new URL(url, location.origin)` instead of string-concatenating `location.origin`, so a path base (`/nuxeo/ui/`) and an absolute base both yield a correct absolute URL — an absolute `base-url` no longer double-prefixes and breaks every restore. Kept an explicit `target.origin === location.origin` same-origin check (the real open-redirect guard, against the window we navigate) plus the under-base `startsWith` and the not-current-page check.
- **Non-string payload:** added a `typeof requestedUrl !== 'string'` early-return before any `startsWith()`, so a tampered `{ "url": 1 }` is ignored rather than throwing and aborting the `_observeCurrentUser` wiring. `URL()` construction is also wrapped so an unparseable value can't throw.

Added tests: restore works with a same-origin absolute `baseUrl`, and a non-string `url` is ignored without throwing (key consumed, no redirect).

File: elements/behaviors/nuxeo-inactivity-behavior.js [hunk 2]
```diff
@@ -272,6 +283,74 @@ export const NuxeoInactivityBehavior = {
...
+      return;
+    }
+    // Restrict to the UI base path (e.g. `/nuxeo/ui/`, which implies same-origin) so a saved value can't
+    // slip past into an open redirect. baseUrl is absent in tests/edge cases → fall back to a bare '/'.
+    // Normalize to a single trailing '/' (mirroring the origin `+ '/'` reasoning): without it a baseUrl of
+    // `/nuxeo/ui` would let a look-alike sibling path such as `${origin}/nuxeo/ui-foo/...` pass startsWith().
+    const baseUrl = this.baseUrl || '/';
+    const normalizedBase = baseUrl.endsWith('/') ? baseUrl : `${baseUrl}/`;
+    const base = `${globalThis.location.origin}${normalizedBase}`;
+    if (requestedUrl?.startsWith(base) && requestedUrl !== globalThis.location.href) {
```
[the remark this answers is not in the document]
Comment (narasimhahyland) [reply elements/behaviors/nuxeo-inactivity-behavior.js]: Fixed in lts-2025 `592ec930f` (cherry-picked to maintenance-3.1.x as `42a70800e`). Added a `typeof requestedUrl !== 'string'` guard before `startsWith`, and wrapped `URL()` construction, so a `{ "url": 42 }`-style payload is ignored (key consumed, no navigation) instead of throwing and aborting `_observeCurrentUser`. Test added.

File: elements/behaviors/nuxeo-inactivity-behavior.js [hunk 3]
```diff
@@ -272,6 +283,74 @@ export const NuxeoInactivityBehavior = {
...
+    if (savedUser !== currentUser.id) {
+      return;
+    }
+    // Restrict to the UI base path (e.g. `/nuxeo/ui/`, which implies same-origin) so a saved value can't
+    // slip past into an open redirect. baseUrl is absent in tests/edge cases → fall back to a bare '/'.
+    // Normalize to a single trailing '/' (mirroring the origin `+ '/'` reasoning): without it a baseUrl of
+    // `/nuxeo/ui` would let a look-alike sibling path such as `${origin}/nuxeo/ui-foo/...` pass startsWith().
+    const baseUrl = this.baseUrl || '/';
+    const normalizedBase = baseUrl.endsWith('/') ? baseUrl : `${baseUrl}/`;
+    const base = `${globalThis.location.origin}${normalizedBase}`;
```
[the remark this answers is not in the document]
Comment (narasimhahyland) [reply elements/behaviors/nuxeo-inactivity-behavior.js]: Fixed in lts-2025 `592ec930f` (cherry-picked to maintenance-3.1.x as `42a70800e`). The base and saved URL are now resolved with `new URL(x, location.origin)` instead of string-concatenating the origin, with an explicit `target.origin === location.origin` same-origin check retained. A same-origin absolute `baseUrl` now restores correctly; test added.

File: elements/behaviors/nuxeo-inactivity-behavior.js [hunk 4]
```diff
@@ -272,6 +283,74 @@ export const NuxeoInactivityBehavior = {
...
   },
 
+  // WEBUI-2189: persist the current page (full hashbang URL) before an inactivity/401 logout navigation,
+  // so _restoreRequestedUrlAfterLogin() can return the user here once they log back in. The owning user's
+  // id is stored alongside the URL so restore only ever returns a page to the same user who left it.
+  _saveRequestedUrl() {
+    try {
+      const payload = JSON.stringify({
+        url: globalThis.location.href,
+        user: this.currentUser?.id,
```
[the remark this answers is not in the document]
Comment (narasimhahyland) [reply elements/behaviors/nuxeo-inactivity-behavior.js]: Good edge case — fixed in lts-2025 `3f1c5b48f` (cherry-picked to maintenance-3.1.x as `9a2f9fbb6`). `_saveRequestedUrl()` now returns early when `this.currentUser?.id` is absent, so we never write an unrestorable `{ user: undefined }` payload that would be consumed but never navigated. Added a test asserting no `sessionStorage.setItem` when there is no resolved user.

[the remark this answers is not in the document]
Comment (narasimhahyland) [reply elements/behaviors/nuxeo-inactivity-behavior.js]: Good edge case — fixed in lts-2025 `3f1c5b48f` (cherry-picked to maintenance-3.1.x as `9a2f9fbb6`). `_saveRequestedUrl()` now returns early when `this.currentUser?.id` is absent, so we never write an unrestorable `{ user: undefined }` payload that would be consumed but never navigated. Added a test asserting no `sessionStorage.setItem` when there is no resolved user.

File: elements/nuxeo-app.js [hunk 5]
```diff
@@ -1509,6 +1509,12 @@ Polymer({
 
   _observeCurrentUser() {
     if (this.currentUser) {
+      // WEBUI-2189: run the post-login restore here rather than from ready(): with anonymous auth enabled
+      // the app boots as Guest before a real user resolves, and restoring from ready() would consume the
+      // saved page for the wrong (or no) user. By the time a real currentUser resolves the router is
+      // listening, so navigating to the saved deep link works. The restore method itself skips anonymous
+      // users and only navigates for the user who saved the page.
+      this._restoreRequestedUrlAfterLogin();
```
[the remark this answers is not in the document]
Comment (narasimhahyland) [reply elements/nuxeo-app.js:1517]: Agreed, and thanks for confirming it is safe — the key is consume-once so repeat observer fires are cheap no-ops, hence no code change. Restore is intentionally driven off the `currentUser` observer (rather than a one-shot) so it also covers the anonymous→real-user boot transition; leaving as-is for that reason.

**Review by vivekchaturved013 — APPROVED** [review]

**Review by naveen-konda — APPROVED** [review]

**Review by madhurkulshrestha-hyland — APPROVED** [review]
